File: players.js
const INITIAL_PLAYERS = [
  // --- GOALKEEPERS (GK) ---
  { id: "gk1", name: "Thibaut Courtois", position: "GK", rating: 90, club: "Real Madrid", nationality: "Belgium", basePrice: 45 },
  { id: "gk2", name: "Alisson Becker", position: "GK", rating: 89, club: "Liverpool", nationality: "Brazil", basePrice: 40 },
  { id: "gk3", name: "Ederson Moraes", position: "GK", rating: 88, club: "Manchester City", nationality: "Brazil", basePrice: 38 },
  { id: "gk4", name: "Marc-André ter Stegen", position: "GK", rating: 89, club: "Barcelona", nationality: "Germany", basePrice: 35 },
  { id: "gk5", name: "Mike Maignan", position: "GK", rating: 87, club: "AC Milan", nationality: "France", basePrice: 35 },
  { id: "gk6", name: "Jan Oblak", position: "GK", rating: 87, club: "Atlético Madrid", nationality: "Slovenia", basePrice: 30 },
  { id: "gk7", name: "Gianluigi Donnarumma", position: "GK", rating: 87, club: "PSG", nationality: "Italy", basePrice: 40 },
  { id: "gk8", name: "Emiliano Martínez", position: "GK", rating: 86, club: "Aston Villa", nationality: "Argentina", basePrice: 28 },
  { id: "gk9", name: "Manuel Neuer", position: "GK", rating: 85, club: "Bayern Munich", nationality: "Germany", basePrice: 15 },
  { id: "gk10", name: "Yann Sommer", position: "GK", rating: 84, club: "Inter Milan", nationality: "Switzerland", basePrice: 12 },
  { id: "gk11", name: "André Onana", position: "GK", rating: 84, club: "Manchester United", nationality: "Cameroon", basePrice: 25 },
  { id: "gk12", name: "Guglielmo Vicario", position: "GK", rating: 83, club: "Tottenham", nationality: "Italy", basePrice: 22 },
  { id: "gk13", name: "Jordan Pickford", position: "GK", rating: 83, club: "Everton", nationality: "England", basePrice: 18 },

  // --- DEFENDERS (DF) ---
  { id: "df1", name: "Virgil van Dijk", position: "DF", rating: 89, club: "Liverpool", nationality: "Netherlands", basePrice: 50 },
  { id: "df2", name: "Rúben Dias", position: "DF", rating: 89, club: "Manchester City", nationality: "Portugal", basePrice: 55 },
  { id: "df3", name: "William Saliba", position: "DF", rating: 88, club: "Arsenal", nationality: "France", basePrice: 60 },
  { id: "df4", name: "John Stones", position: "DF", rating: 87, club: "Manchester City", nationality: "England", basePrice: 42 },
  { id: "df5", name: "Ronald Araújo", position: "DF", rating: 86, club: "Barcelona", nationality: "Uruguay", basePrice: 50 },
  { id: "df6", name: "Marquinhos", position: "DF", rating: 87, club: "PSG", nationality: "Brazil", basePrice: 40 },
  { id: "df7", name: "Éder Militão", position: "DF", rating: 86, club: "Real Madrid", nationality: "Brazil", basePrice: 48 },
  { id: "df8", name: "Antonio Rüdiger", position: "DF", rating: 87, club: "Real Madrid", nationality: "Germany", basePrice: 38 },
  { id: "df9", name: "Gabriel Magalhães", position: "DF", rating: 86, club: "Arsenal", nationality: "Brazil", basePrice: 45 },
  { id: "df10", name: "Alessandro Bastoni", position: "DF", rating: 86, club: "Inter Milan", nationality: "Italy", basePrice: 48 },
  { id: "df11", name: "Trent Alexander-Arnold", position: "DF", rating: 86, club: "Liverpool", nationality: "England", basePrice: 55 },
  { id: "df12", name: "Achraf Hakimi", position: "DF", rating: 86, club: "PSG", nationality: "Morocco", basePrice: 50 },
  { id: "df13", name: "Theo Hernández", position: "DF", rating: 86, club: "AC Milan", nationality: "France", basePrice: 48 },
  { id: "df14", name: "Alphonso Davies", position: "DF", rating: 84, club: "Bayern Munich", nationality: "Canada", basePrice: 42 },
  { id: "df15", name: "Kyle Walker", position: "DF", rating: 84, club: "Manchester City", nationality: "England", basePrice: 20 },
  { id: "df16", name: "Dani Carvajal", position: "DF", rating: 85, club: "Real Madrid", nationality: "Spain", basePrice: 22 },
  { id: "df17", name: "Jules Koundé", position: "DF", rating: 85, club: "Barcelona", nationality: "France", basePrice: 40 },
  { id: "df18", name: "Josko Gvardiol", position: "DF", rating: 85, club: "Manchester City", nationality: "Croatia", basePrice: 55 },
  { id: "df19", name: "Jeremie Frimpong", position: "DF", rating: 84, club: "Bayer Leverkusen", nationality: "Netherlands", basePrice: 45 },
  { id: "df20", name: "Cristian Romero", position: "DF", rating: 85, club: "Tottenham", nationality: "Argentina", basePrice: 45 },
  { id: "df21", name: "Kim Min-jae", position: "DF", rating: 84, club: "Bayern Munich", nationality: "South Korea", basePrice: 40 },
  { id: "df22", name: "Matthijs de Ligt", position: "DF", rating: 84, club: "Manchester United", nationality: "Netherlands", basePrice: 38 },
  { id: "df23", name: "Kieran Trippier", position: "DF", rating: 83, club: "Newcastle", nationality: "England", basePrice: 12 },
  { id: "df24", name: "Andrew Robertson", position: "DF", rating: 84, club: "Liverpool", nationality: "Scotland", basePrice: 28 },
  { id: "df25", name: "Nico Schlotterbeck", position: "DF", rating: 83, club: "Borussia Dortmund", nationality: "Germany", basePrice: 30 },
  { id: "df26", name: "Fikayo Tomori", position: "DF", rating: 83, club: "AC Milan", nationality: "England", basePrice: 28 },
  { id: "df27", name: "Gleison Bremer", position: "DF", rating: 84, club: "Juventus", nationality: "Brazil", basePrice: 35 },
  { id: "df28", name: "Ben White", position: "DF", rating: 83, club: "Arsenal", nationality: "England", basePrice: 35 },
  { id: "df29", name: "Federico Dimarco", position: "DF", rating: 84, club: "Inter Milan", nationality: "Italy", basePrice: 38 },
  { id: "df30", name: "Pau Torres", position: "DF", rating: 83, club: "Aston Villa", nationality: "Spain", basePrice: 28 },

  // --- MIDFIELDERS (MF) ---
  { id: "mf1", name: "Kevin De Bruyne", position: "MF", rating: 91, club: "Manchester City", nationality: "Belgium", basePrice: 65 },
  { id: "mf2", name: "Rodri (Rodrigo Hernández)", position: "MF", rating: 90, club: "Manchester City", nationality: "Spain", basePrice: 85 },
  { id: "mf3", name: "Jude Bellingham", position: "MF", rating: 90, club: "Real Madrid", nationality: "England", basePrice: 95 },
  { id: "mf4", name: "Declan Rice", position: "MF", rating: 88, club: "Arsenal", nationality: "England", basePrice: 70 },
  { id: "mf5", name: "Martin Ødegaard", position: "MF", rating: 89, club: "Arsenal", nationality: "Norway", basePrice: 75 },
  { id: "mf6", name: "Jamal Musiala", position: "MF", rating: 88, club: "Bayern Munich", nationality: "Germany", basePrice: 80 },
  { id: "mf7", name: "Florian Wirtz", position: "MF", rating: 88, club: "Bayer Leverkusen", nationality: "Germany", basePrice: 80 },
  { id: "mf8", name: "Pedri González", position: "MF", rating: 86, club: "Barcelona", nationality: "Spain", basePrice: 60 },
  { id: "mf9", name: "Federico Valverde", position: "MF", rating: 88, club: "Real Madrid", nationality: "Uruguay", basePrice: 75 },
  { id: "mf10", name: "Bruno Fernandes", position: "MF", rating: 87, club: "Manchester United", nationality: "Portugal", basePrice: 50 },
  { id: "mf11", name: "Ilkay Gündogan", position: "MF", rating: 86, club: "Barcelona", nationality: "Germany", basePrice: 20 },
  { id: "mf12", name: "Bernardo Silva", position: "MF", rating: 88, club: "Manchester City", nationality: "Portugal", basePrice: 55 },
  { id: "mf13", name: "Nicolò Barella", position: "MF", rating: 87, club: "Inter Milan", nationality: "Italy", basePrice: 60 },
  { id: "mf14", name: "Hakan Çalhanoglu", position: "MF", rating: 86, club: "Inter Milan", nationality: "Turkey", basePrice: 40 },
  { id: "mf15", name: "Luka Modric", position: "MF", rating: 85, club: "Real Madrid", nationality: "Croatia", basePrice: 10 },
  { id: "mf16", name: "Alexis Mac Allister", position: "MF", rating: 85, club: "Liverpool", nationality: "Argentina", basePrice: 55 },
  { id: "mf17", name: "Dominik Szoboszlai", position: "MF", rating: 84, club: "Liverpool", nationality: "Hungary", basePrice: 50 },
  { id: "mf18", name: "Aurelien Tchouaméni", position: "MF", rating: 85, club: "Real Madrid", nationality: "France", basePrice: 55 },
  { id: "mf19", name: "Eduardo Camavinga", position: "MF", rating: 85, club: "Real Madrid", nationality: "France", basePrice: 60 },
  { id: "mf20", name: "Joshua Kimmich", position: "MF", rating: 86, club: "Bayern Munich", nationality: "Germany", basePrice: 45 },
  { id: "mf21", name: "Enzo Fernández", position: "MF", rating: 84, club: "Chelsea", nationality: "Argentina", basePrice: 50 },
  { id: "mf22", name: "Bruno Guimarães", position: "MF", rating: 85, club: "Newcastle", nationality: "Brazil", basePrice: 55 },
  { id: "mf23", name: "James Maddison", position: "MF", rating: 84, club: "Tottenham", nationality: "England", basePrice: 38 },
  { id: "mf24", name: "Frenkie de Jong", position: "MF", rating: 86, club: "Barcelona", nationality: "Netherlands", basePrice: 48 },
  { id: "mf25", name: "Gavi (Pablo Martín)", position: "MF", rating: 83, club: "Barcelona", nationality: "Spain", basePrice: 50 },
  { id: "mf26", name: "Granit Xhaka", position: "MF", rating: 84, club: "Bayer Leverkusen", nationality: "Switzerland", basePrice: 20 },
  { id: "mf27", name: "Warren Zaïre-Emery", position: "MF", rating: 82, club: "PSG", nationality: "France", basePrice: 45 },
  { id: "mf28", name: "Douglas Luiz", position: "MF", rating: 84, club: "Juventus", nationality: "Brazil", basePrice: 38 },
  { id: "mf29", name: "Teun Koopmeiners", position: "MF", rating: 83, club: "Juventus", nationality: "Netherlands", basePrice: 35 },
  { id: "mf30", name: "João Palhinha", position: "MF", rating: 84, club: "Bayern Munich", nationality: "Portugal", basePrice: 35 },

  // --- FORWARDS (FW) ---
  { id: "fw1", name: "Kylian Mbappé", position: "FW", rating: 91, club: "Real Madrid", nationality: "France", basePrice: 110 },
  { id: "fw2", name: "Erling Haaland", position: "FW", rating: 91, club: "Manchester City", nationality: "Norway", basePrice: 110 },
  { id: "fw3", name: "Harry Kane", position: "FW", rating: 90, club: "Bayern Munich", nationality: "England", basePrice: 75 },
  { id: "fw4", name: "Vinícius Júnior", position: "FW", rating: 90, club: "Real Madrid", nationality: "Brazil", basePrice: 100 },
  { id: "fw5", name: "Mohamed Salah", position: "FW", rating: 89, club: "Liverpool", nationality: "Egypt", basePrice: 65 },
  { id: "fw6", name: "Bukayo Saka", position: "FW", rating: 89, club: "Arsenal", nationality: "England", basePrice: 85 },
  { id: "fw7", name: "Antoine Griezmann", position: "FW", rating: 88, club: "Atlético Madrid", nationality: "France", basePrice: 35 },
  { id: "fw8", name: "Phil Foden", position: "FW", rating: 89, club: "Manchester City", nationality: "England", basePrice: 85 },
  { id: "fw9", name: "Son Heung-min", position: "FW", rating: 87, club: "Tottenham", nationality: "South Korea", basePrice: 40 },
  { id: "fw10", name: "Robert Lewandowski", position: "FW", rating: 88, club: "Barcelona", nationality: "Poland", basePrice: 25 },
  { id: "fw11", name: "Lautaro Martínez", position: "FW", rating: 89, club: "Inter Milan", nationality: "Argentina", basePrice: 70 },
  { id: "fw12", name: "Victor Osimhen", position: "FW", rating: 88, club: "Galatasaray", nationality: "Nigeria", basePrice: 65 },
  { id: "fw13", name: "Rafael Leão", position: "FW", rating: 86, club: "AC Milan", nationality: "Portugal", basePrice: 60 },
  { id: "fw14", name: "Rodrygo Goes", position: "FW", rating: 86, club: "Real Madrid", nationality: "Brazil", basePrice: 65 },
  { id: "fw15", name: "(GOAT) Lionel Messi", position: "FW", rating: 88, club: "Inter Miami", nationality: "Argentina", basePrice: 20 },
  { id: "fw16", name: "Cristiano Ronaldo", position: "FW", rating: 86, club: "Al Nassr", nationality: "Portugal", basePrice: 15 },
  { id: "fw17", name: "Neymar Jr", position: "FW", rating: 86, club: "Al Hilal", nationality: "Brazil", basePrice: 20 },
  { id: "fw18", name: "Luis Díaz", position: "FW", rating: 84, club: "Liverpool", nationality: "Colombia", basePrice: 48 },
  { id: "fw19", name: "Marcus Rashford", position: "FW", rating: 84, club: "Manchester United", nationality: "England", basePrice: 40 },
  { id: "fw20", name: "Gabriel Martinelli", position: "FW", rating: 84, club: "Arsenal", nationality: "Brazil", basePrice: 45 },
  { id: "fw21", name: "Khvicha Kvaratskhelia", position: "FW", rating: 85, club: "Napoli", nationality: "Georgia", basePrice: 55 },
  { id: "fw22", name: "Ousmane Dembélé", position: "FW", rating: 85, club: "PSG", nationality: "France", basePrice: 42 },
  { id: "fw23", name: "Leroy Sané", position: "FW", rating: 84, club: "Bayern Munich", nationality: "Germany", basePrice: 40 },
  { id: "fw24", name: "Julián Álvarez", position: "FW", rating: 84, club: "Atlético Madrid", nationality: "Argentina", basePrice: 55 },
  { id: "fw25", name: "Darwin Núñez", position: "FW", rating: 82, club: "Liverpool", nationality: "Uruguay", basePrice: 35 },
  { id: "fw26", name: "Alexander Isak", position: "FW", rating: 84, club: "Newcastle", nationality: "Sweden", basePrice: 55 },
  { id: "fw27", name: "Ollie Watkins", position: "FW", rating: 84, club: "Aston Villa", nationality: "England", basePrice: 48 },
  { id: "fw28", name: "Dušan Vlahović", position: "FW", rating: 84, club: "Juventus", nationality: "Serbia", basePrice: 45 },
  { id: "fw29", name: "Viktor Gyökeres", position: "FW", rating: 83, club: "Sporting CP", nationality: "Sweden", basePrice: 45 },
  { id: "fw30", name: "Cole Palmer", position: "FW", rating: 86, club: "Chelsea", nationality: "England", basePrice: 70 },
  { id: "fw31", name: "Lamine Yamal", position: "FW", rating: 84, club: "Barcelona", nationality: "Spain", basePrice: 75 },
  { id: "fw32", name: "Nico Williams", position: "FW", rating: 84, club: "Athletic Bilbao", nationality: "Spain", basePrice: 50 },

  // --- INDIAN PLAYERS ---
  { id: "ind1", name: "Sunil Chhetri", position: "FW", rating: 75, club: "Bengaluru FC", nationality: "India", basePrice: 5 },
  { id: "ind2", name: "Gurpreet Singh Sandhu", position: "GK", rating: 72, club: "Bengaluru FC", nationality: "India", basePrice: 3 },
  { id: "ind3", name: "Sandesh Jhingan", position: "DF", rating: 73, club: "FC Goa", nationality: "India", basePrice: 3 },
  { id: "ind4", name: "Lallianzuala Chhangte", position: "FW", rating: 74, club: "Mumbai City FC", nationality: "India", basePrice: 4 },
  { id: "ind5", name: "Anirudh Thapa", position: "MF", rating: 71, club: "Mohun Bagan SG", nationality: "India", basePrice: 3 },
  { id: "ind6", name: "Liston Colaco", position: "FW", rating: 72, club: "Mohun Bagan SG", nationality: "India", basePrice: 3 },
  { id: "ind7", name: "Subhasish Bose", position: "DF", rating: 70, club: "Mohun Bagan SG", nationality: "India", basePrice: 2 },
  { id: "ind8", name: "Sahal Abdul Samad", position: "MF", rating: 72, club: "Mohun Bagan SG", nationality: "India", basePrice: 3 },
  { id: "ind9", name: "Manvir Singh", position: "FW", rating: 71, club: "Mohun Bagan SG", nationality: "India", basePrice: 3 },
  { id: "ind10", name: "Anwar Ali", position: "DF", rating: 71, club: "East Bengal FC", nationality: "India", basePrice: 2 },
  { id: "ind11", name: "Lalengmawia Ralte", position: "MF", rating: 72, club: "Mohun Bagan SG", nationality: "India", basePrice: 3 },
  { id: "ind12", name: "Jeakson Singh", position: "MF", rating: 70, club: "East Bengal FC", nationality: "India", basePrice: 2 },
  { id: "ind13", name: "Vishal Kaith", position: "GK", rating: 71, club: "Mohun Bagan SG", nationality: "India", basePrice: 2 },
  { id: "ind14", name: "Mehtab Singh", position: "DF", rating: 69, club: "Mumbai City FC", nationality: "India", basePrice: 2 },
  { id: "ind15", name: "Naorem Mahesh Singh", position: "FW", rating: 72, club: "East Bengal FC", nationality: "India", basePrice: 3 }
];

// Helper to retrieve players dynamically without localStorage
function getPlayersDatabase() {
  const combined = [...INITIAL_PLAYERS];
  const existingNames = new Set(combined.map(p => p.name.toLowerCase().trim()));
  const existingIds = new Set(combined.map(p => p.id));

  if (typeof WC2026_PLAYERS !== 'undefined') {
    WC2026_PLAYERS.forEach(p => {
      if (!existingIds.has(p.id) && !existingNames.has(p.name.toLowerCase().trim())) {
        combined.push(p);
        existingIds.add(p.id);
        existingNames.add(p.name.toLowerCase().trim());
      }
    });
  }
  return combined;
}

function getPlayerCareerFantasyPoints(p) {
  const nameLower = (p.name || '').toLowerCase().trim();
  if (nameLower.includes("messi")) return 990;
  if (nameLower.includes("ronaldo")) return 980;
  if (nameLower.includes("mbappé") || nameLower.includes("mbappe")) return 920;
  if (nameLower.includes("haaland")) return 895;
  if (nameLower.includes("kane")) return 830;
  if (nameLower.includes("bellingham")) return 880;
  if (nameLower.includes("chhetri")) return 760;
  if (nameLower.includes("de bruyne")) return 910;
  if (nameLower.includes("neymar")) return 890;
  if (nameLower.includes("salah")) return 880;
  if (nameLower.includes("modric")) return 940;
  if (nameLower.includes("kroos")) return 950;
  if (nameLower.includes("lewandowski")) return 900;
  if (nameLower.includes("courtois")) return 880;
  if (nameLower.includes("van dijk")) return 870;

  const base = Math.round(((p.rating || 80) / 99) * 700);

  let clubPoints = 30;
  const clubLower = (p.club || '').toLowerCase();
  if (clubLower.includes("real madrid") || clubLower.includes("manchester city") || clubLower.includes("bayern") || clubLower.includes("barcelona") || clubLower.includes("liverpool")) {
    clubPoints = 150;
  } else if (clubLower.includes("inter") || clubLower.includes("milan") || clubLower.includes("psg") || clubLower.includes("juventus") || clubLower.includes("arsenal") || clubLower.includes("atlético") || clubLower.includes("atletico") || clubLower.includes("dormund")) {
    clubPoints = 110;
  } else if (clubLower.includes("tottenham") || clubLower.includes("chelsea") || clubLower.includes("manchester united") || clubLower.includes("bayer leverkusen") || clubLower.includes("aston villa") || clubLower.includes("newcastle") || clubLower.includes("bengaluru") || clubLower.includes("mohun bagan") || clubLower.includes("east bengal") || clubLower.includes("goa")) {
    clubPoints = 70;
  }

  let nationPoints = 40;
  const nationLower = (p.nationality || '').toLowerCase();
  if (nationLower.includes("argentina") || nationLower.includes("france") || nationLower.includes("spain") || nationLower.includes("brazil") || nationLower.includes("germany") || nationLower.includes("italy")) {
    nationPoints = 150;
  } else if (nationLower.includes("portugal") || nationLower.includes("uruguay") || nationLower.includes("england") || nationLower.includes("belgium") || nationLower.includes("netherlands") || nationLower.includes("croatia") || nationLower.includes("colombia")) {
    nationPoints = 100;
  } else if (nationLower.includes("india")) {
    nationPoints = 65;
  }

  return Math.min(1000, base + clubPoints + nationPoints);
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { INITIAL_PLAYERS, getPlayersDatabase, getPlayerCareerFantasyPoints };
}

File: players.py
# ============================================================
# players.py - Player Database & Performance Form Service (Python)
# ============================================================
import re
import json
import os
from typing import List, Dict, Any, Optional

def load_initial_players(file_path: str = "public/players.js") -> List[Dict[str, Any]]:
    """Loads the main initial players list from players.js by parsing the JSON array."""
    if not os.path.exists(file_path):
        return []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        # Find the content between const INITIAL_PLAYERS = [ and ];
        match = re.search(r"const\s+INITIAL_PLAYERS\s*=\s*(\[.*?\]);", content, re.DOTALL)
        if match:
            return json.loads(match.group(1))
    except Exception as e:
        print(f"Error loading players.js: {e}")
    return []


def load_wc2026_players(file_path: str = "public/wc2026_players.js") -> List[Dict[str, Any]]:
    """Loads the World Cup players list from wc2026_players.js."""
    if not os.path.exists(file_path):
        return []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        # Find the content between const WC2026_PLAYERS = [ and ];
        match = re.search(r"const\s+WC2026_PLAYERS\s*=\s*(\[.*?\]);", content, re.DOTALL)
        if match:
            return json.loads(match.group(1))
    except Exception as e:
        print(f"Error loading wc2026_players.js: {e}")
    return []


def get_js_bitwise_hash(name: str) -> int:
    """
    Simulates the JavaScript string hashing algorithm used in players.js:
    for (let i = 0; i < nameLower.length; i++) {
        hash = (hash << 5) - hash + nameLower.charCodeAt(i);
        hash |= 0;
    }
    This ensures identical deterministic seed generation and form statuses.
    """
    name_lower = name.lower()
    hash_val = 0
    for char in name_lower:
        # (hash_val << 5) - hash_val + ord(char)
        hash_val = (hash_val << 5) - hash_val + ord(char)
        # Emulate 32-bit bitwise operations in JavaScript
        hash_val = int(hash_val) & 0xFFFFFFFF
        if hash_val >= 0x80000000:
            hash_val -= 0x100000000
    
    return abs(hash_val) % 100


def apply_present_performance(p: Dict[str, Any]) -> Dict[str, Any]:
    """
    Calculates form boosts, rating shifts, and updates player base prices dynamically
    exactly like the JavaScript applyPresentPerformance(p) function.
    """
    if not p:
        return p
    if p.get("_performanceApplied"):
        return p

    base_rating = p.get("rating", 75)
    name = p.get("name", "")
    name_lower = name.lower()

    form_boost = 0
    form_status = '➖ Steady'
    form_class = 'steady'

    seed = get_js_bitwise_hash(name)

    # Check key fire-players
    fire_names = ['mbappé', 'mbappe', 'bellingham', 'haaland', 'vinícius', 'yamal', 'rodri', 'kane']
    if any(fn in name_lower for fn in fire_names):
        form_boost = 3
        form_status = '🔥 On Fire'
        form_class = 'on-fire'
    elif seed > 75:
        form_boost = 3
        form_status = '🔥 On Fire'
        form_class = 'on-fire'
    elif seed > 45:
        form_boost = 2
        form_status = '📈 Peak Form'
        form_class = 'peak-form'
    elif seed > 20:
        form_boost = 1
        form_status = '⚡ Good Form'
        form_class = 'good-form'
    else:
        form_boost = 0
        form_status = '➖ Steady'
        form_class = 'steady'

    effective_rating = min(99, max(60, base_rating + form_boost))

    # Base price calculation helper
    def calc_price(r: int) -> int:
        if r >= 90:
            return 50 + (r - 90) * 15
        if r >= 80:
            return 20 + (r - 80) * 3
        if r >= 70:
            return 5 + (r - 70) * 1.5
        return max(1, round(r / 15))

    effective_base_price = calc_price(effective_rating)

    # Modify copy or direct reference attributes
    p["baseRating"] = base_rating
    p["rating"] = effective_rating
    p["formBoost"] = form_boost
    p["formStatus"] = form_status
    p["formClass"] = form_class
    p["basePrice"] = effective_base_price
    p["_performanceApplied"] = True
    return p


def get_players_database(mode: str = 'special') -> List[Dict[str, Any]]:
    """Returns the requested player pool with form boosts applied."""
    raw_list = []
    if mode == 'wc2026_elite':
        wc_players = load_wc2026_players()
        raw_list = [p for p in wc_players if p.get("rating", 0) >= 80]
    elif mode in ('wc2026', 'wc2026_all'):
        raw_list = load_wc2026_players()
    else:
        raw_list = load_initial_players()
        
    return [apply_present_performance(dict(p)) for p in raw_list]


def get_player_career_fantasy_points(p: Dict[str, Any]) -> int:
    """Calculates career points based on player ratings, club tiers, and nationalities."""
    name_lower = p.get("name", "").lower().strip()
    
    # Specific legendary point overrides
    if "messi" in name_lower: return 990
    if "ronaldo" in name_lower: return 980
    if "mbappé" in name_lower or "mbappe" in name_lower: return 920
    if "haaland" in name_lower: return 895
    if "kane" in name_lower: return 830
    if "bellingham" in name_lower: return 880
    if "chhetri" in name_lower: return 760
    if "de bruyne" in name_lower: return 910
    if "neymar" in name_lower: return 890
    if "salah" in name_lower: return 880
    if "modric" in name_lower: return 940
    if "kroos" in name_lower: return 950
    if "lewandowski" in name_lower: return 900
    if "courtois" in name_lower: return 880
    if "van dijk" in name_lower: return 870

    base = round((p.get("rating", 80) / 99) * 700)

    club_points = 30
    club_lower = p.get("club", "").lower()
    elite_clubs = ["real madrid", "manchester city", "bayern", "barcelona", "liverpool"]
    tier2_clubs = ["inter", "milan", "psg", "juventus", "arsenal", "atlético", "atletico", "dormund"]
    tier3_clubs = ["tottenham", "chelsea", "manchester united", "bayer leverkusen", "aston villa", "newcastle", "bengaluru", "mohun bagan", "east bengal", "goa"]
    
    if any(ec in club_lower for ec in elite_clubs):
        club_points = 150
    elif any(t2 in club_lower for t2 in tier2_clubs):
        club_points = 110
    elif any(t3 in club_lower for t3 in tier3_clubs):
        club_points = 70

    nation_points = 40
    nation_lower = p.get("nationality", "").lower()
    elite_nations = ["argentina", "france", "spain", "brazil", "germany", "italy"]
    tier2_nations = ["portugal", "uruguay", "england", "belgium", "netherlands", "croatia", "colombia"]
    
    if any(en in nation_lower for en in elite_nations):
        nation_points = 150
    elif any(t2n in nation_lower for t2n in tier2_nations):
        nation_points = 100
    elif "india" in nation_lower:
        nation_points = 65

    return min(1000, base + club_points + nation_points)


# Preload database
INITIAL_PLAYERS = load_initial_players()
WC2026_PLAYERS = load_wc2026_players()
